src/components/tooltip/tooltip.js

```diff
@@ -10,6 +36,116 @@ import {h, Component} from 'preact';
  * @extends {Component}
  */
 class Tooltip extends Component {
+  state: Object;
+  _hoverTimeout: ?number;
+  textElement: HTMLSpanElement;
+  nonValidTypes: string[] = [];
+
+  /**
+   * default component props
+   * @type {Object}
+   * @memberof Tooltip
+   */
+  static defaultProps = {
+    type: ToolTipType.Top,
+    isMultiline: true,
+    maxWidth: '240px'
+  };
+
+  /**
+   * @constructor
+   * @param {*} props props
+   * @memberof Tooltip
+   */
+  constructor(props: Object) {
+    super(props);
+  }
+
+  /**
+   * on mouse over handler.
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  onMouseOver(): void {
+    if (this._hoverTimeout) {
+      clearTimeout(this._hoverTimeout);
+      this._hoverTimeout = null;
+    }
+    this._hoverTimeout = setTimeout(() => {
+      this.setState({showTooltip: true});
+    }, TOOLTIP_SHOW_TIMEOUT);
+  }
+
+  /**
+   * on mouse leave handler.
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  onMouseLeave(): void {
+    this.setState({showTooltip: false});
+    clearTimeout(this._hoverTimeout);
+    this._hoverTimeout = null;
+  }
+
+  /**
+   * brings another tooltip type which hasn't been marked as invalid
+   * @memberof Tooltip
+   * @returns {string} tooltip type
+   */
+  getAlternateType(): string {
+    for (const key in ToolTipType) {
+      const toolTipType = ToolTipType[key];
+      if (this.nonValidTypes.indexOf(toolTipType) === -1) {
+        return toolTipType;
+      }
+    }
+    return this.state.type;
+  }
+
+  /**
+   * checks if the current tooltip type is within the player boundaries
+   * @memberof Tooltip
+   * @returns {string} is in boundaries
+   */
+  isToolTipInBoundaries(): boolean {
+    const tooltipBoundingRect = this.textElement.getBoundingClientRect();
+    const playerContainerRect = this.props.playerClientRect;
+
+    return (
+      tooltipBoundingRect.top > playerContainerRect.top + PLAYER_MARGIN &&
+      tooltipBoundingRect.bottom < playerContainerRect.bottom - PLAYER_MARGIN &&
+      tooltipBoundingRect.right < playerContainerRect.right - PLAYER_MARGIN &&
+      tooltipBoundingRect.left > playerContainerRect.left + PLAYER_MARGIN
+    );
+  }
+
+  /**
+   * sets the requested type prop of the tooltip as a state cause it can change if is not valid
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentWillMount(): void {
+    this.setState({type: this.props.type});
+  }
+
+  /**
+   * checks if after the render the tooltip is within boundaries of the player
+   * if not it will try to set a new type which will be checked after the next render
+   * @memberof Tooltip
+   * @returns {void}
+   */
+  componentDidUpdate(): void {
+    if (this.state.showTooltip && !this.isToolTipInBoundaries()) {
```

Comment:
Don't you can test the all possible tooltips and select the one which can be used by order?
Here you make array of non valid type but the playerClientRect can change and they will be relevant again

Reply:
fixed